config/300-awss3.yaml

```diff
@@ -0,0 +1,213 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: awss3sources.sources.triggermesh.io
+  labels:
+    eventing.knative.dev/source: 'true'
+    duck.knative.dev/source: 'true'
+    knative.dev/crd-install: 'true'
+  annotations:
+    registry.knative.dev/eventTypes: |
+      [
+        { "type": "com.amazon.s3.objectcreated" },
+        { "type": "com.amazon.s3.objectremoved" },
+        { "type": "com.amazon.s3.objectrestore" },
+        { "type": "com.amazon.s3.reducedredundancylostobject" },
+        { "type": "com.amazon.s3.replication" },
+        { "type": "com.amazon.s3.event" }
+      ]
+spec:
+  group: sources.triggermesh.io
+  scope: Namespaced
+  names:
+    kind: AWSS3Source
+    plural: awss3sources
+    categories:
+    - all
+    - knative
+    - eventing
+    - sources
+  versions:
+  - name: v1alpha1
+    served: true
+    storage: true
+    subresources:
+      status: {}
+    schema:
+      openAPIV3Schema:
+        type: object
+        properties:
+          spec:
+            type: object
+            properties:
+              arn:
+                type: string
+                # Bucket naming rules
+                # https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html
+                pattern: ^arn:aws(-cn|-us-gov)?:s3:[a-z]{2}(-gov)?-[a-z]+-\d:\d{12}:[0-9a-z][0-9a-z.-]{2,62}$
+              eventTypes:
+                type: array
+                items:
+                  type: string
+                  # Accepted values
+                  # https://docs.aws.amazon.com/AmazonS3/latest/API/API_QueueConfiguration.html
+                  # https://docs.aws.amazon.com/AmazonS3/latest/userguide/notification-how-to-event-types-and-destinations.html
+                  enum:
+                  - s3:ObjectCreated:*
+                  - s3:ObjectCreated:Put
+                  - s3:ObjectCreated:Post
+                  - s3:ObjectCreated:Copy
+                  - s3:ObjectCreated:CompleteMultipartUpload
+                  - s3:ObjectRemoved:*
+                  - s3:ObjectRemoved:Delete
+                  - s3:ObjectRemoved:DeleteMarkerCreated
+                  - s3:ObjectRestore:*
+                  - s3:ObjectRestore:Post
+                  - s3:ObjectRestore:Completed
+                  - s3:ReducedRedundancyLostObject
+                  - s3:Replication:*
+                  - s3:Replication:OperationFailedReplication
+                  - s3:Replication:OperationNotTracked
+                  - s3:Replication:OperationMissedThreshold
+                  - s3:Replication:OperationReplicatedAfterThreshold
+              queueARN:
```

Comment:
Yes I added it as a fallback because S3 sends a test event with a different schema every time the configuration of event notifications changes in the bucket.

Now that you mention it, I could simply plan for that case and add a "testevent" type. The list of supported events is exhaustive in the spec after all.

Reply:
I wanted to understand that one, maybe `testevent` could be confusing. I think it is better as is for users that want to map to S3 events.

Reply:
I ended up improving the current situation in 742dba9 by
* handling the `com.amazon.s3.testevent` case
* falling back to `com.amazon.sqs.message` when the received message can not be identified as originating from S3

The second case would indicate a flaw in the way customers use their SQS queue, but at least we try to handle things gracefully instead of discarding messages altogether.

While I was at it, I also removed the usage of `ceOverrides` in `adapter.go`, because it was unnecessary and made the origin of the "source" attribute unclear.